src/lamp_py/bus_performance_manager/vehicle_events.py

```diff
@@ -152,3 +155,115 @@ def get_service_date_from_filename(tm_filename: str) -> Optional[date]:
+    """
+    bus_routes = bus_routes_for_service_date(service_date)
+
+    # build a dataframe of every gtfs record
+    # * scan all of the gtfs realtime files
+    # * only pull out bus records and records for the service date
+    # * rename / convert columns as appropriate
+    # * sort by vehicle id and timestamp
+    # * keep only the first record for a given trip / stop / status
+    # * sort again by vehicle id and timestamp (the unique call seems to reorder)
+    vehicle_positions = (
+        pl.scan_parquet(gtfs_rt_files)
+        .filter(
+            (pl.col("vehicle.trip.route_id").is_in(bus_routes))
+            & (
+                pl.col("vehicle.trip.start_date")
+                == service_date.strftime("%Y%m%d")
+            )
+        )
+        .select(
+            pl.col("vehicle.trip.route_id").cast(pl.String).alias("route_id"),
+            pl.col("vehicle.trip.trip_id").cast(pl.String).alias("trip_id"),
+            pl.col("vehicle.stop_id").cast(pl.String).alias("stop_id"),
+            pl.col("vehicle.trip.direction_id").alias("direction_id"),
+            pl.col("vehicle.trip.start_time").alias("start_time"),
+            pl.col("vehicle.trip.start_date").alias("service_date"),
+            pl.col("vehicle.vehicle.id").cast(pl.String).alias("vehicle_id"),
+            pl.col("vehicle.current_status").alias("current_status"),
+            pl.from_epoch("vehicle.timestamp")
+            .dt.convert_time_zone("EST5EDT")
+            .alias("vehicle_timestamp"),
+        )
+        .unique()
+        .sort(["vehicle_id", "vehicle_timestamp"])
+        .unique(
+            subset=[
+                "route_id",
+                "trip_id",
+                "stop_id",
+                "direction_id",
+                "vehicle_id",
+                "current_status",
+            ],
+            keep="first",
+        )
+        .sort(["vehicle_id", "vehicle_timestamp"])
+        .collect()
+    )
+
+    # using the vehicle positions dataframe, create a dataframe for each event
+    # by pivoting and mapping the current status onto arrivals and departures.
+    # initially, they map to arrival times and traveling towards times. shift
+    # the traveling towards datetime over trip id to get the departure time.
+    vehicle_events = (
+        vehicle_positions.pivot(
+            values="vehicle_timestamp",
+            index=[
+                "route_id",
+                "trip_id",
+                "stop_id",
+                "start_time",
+                "service_date",
+                "vehicle_id",
+            ],
+            columns="current_status",
+        )
+        .rename(
+            {
+                "STOPPED_AT": "arrival_gtfs",
+                "IN_TRANSIT_TO": "started_towards_datetime",
+            }
+        )
+        .with_columns(
+            pl.col("started_towards_datetime")
+            .shift(-1)
+            .over("trip_id")
+            .alias("departure_gtfs")
```

Comment:
I would advise against doing any shift operations until after this dataset is joined with Transit  Master data. Since we would presumably want to do the same shift operations against that dataset as well.